# Taller de Manejo y Análisis de Datos

**Profesor**: Pedro Montealegre

# Funciones y Módulos


Las funciones nos permiten agrupar cierto número de operaciones en un solo *bloque lógico*. Nos comunicamos con una función a través de una interfaz claramente definida, entregando ciertos parámetros a la función, y recibiendo cierta información de vuelta. A parte de esta interfaz, por lo general no sabemos exactamente cómo la función es capaz de producir el valor de salida. 

Por ejemplo, cuando escribimos:

In [1]:
import math
math.sqrt(2)

1.4142135623730951

No sabemos exactamente cómo la función `math.sqrt` calcula la raíz cuadrdada de $2$, pero conocemos la interfaz: si le pasamos $x$ a la función, ésta nos entrega (una aproximación de) $\sqrt{x}$. 

Esta *abstracción* es muy útil en programación, y en general se usa en todas las ramas de la ingeniería: se divide un sistema en varias partes más pequeñas, que interactúan a través de interfaces bien definidas, y que no necesitan saber sobre el funcionamiento interno de las demás. De hecho, no necesitar preocuparse de los detalles de estas implementaciones por lo general ayuda a tener una visión más clara de cuáles son las partes críticas del sistema, y así asegurar un buen funcionamiento.

Las funciones son el "bloque de construcción" (building block) sobre el cual se basa la funcionalidad en programas más grandes (y simulaciones computacionales), y ayudan a controlar la complejidad inherente al proceso. 

Se puede agrupar varias funciones en un módulo Python, y de este modo crear nuestras propias librerías.


## Usando funciones

La palabra "función" tiene diferentes significados en matemáticas y en programación. En programación, se refiere a una secuencia de operaciones a las cuales se les asigna un nombre. Por ejemplo, la función `sqrt()`, que está definida en el módulo `math` calcula la raíz cuadrada de un número:

In [2]:
import math
math.sqrt(4)

2.0

En este ejemplo, el valor que le pasamos a la función `sqrt` es 4.  Este valor se llama el *argumento* de la función. Una función puede tener cero, uno, o más de un argumento. 

La función devuelve el valor 2.0 al "contexto de llamada". Este valor se conoce como el *valor de salida* de la función.

Se dice que una función *recibe* un argumento y *devuelve* un valor (o resultado).

#### Confusión entre imprimir y devolver

Un error común entre quienes se están iniciando en la programación consiste en confundir **imprimir** un valor con **devolver** un valor. 

Parte de la confusion sobre valores devueltos versus valores impresos viene por el hecho que el terminal Python imprime (una representación de) los objetos devueltos **siempre y cuando** estos no hayan sido asignados a una variable. Por ejemplo:

In [3]:
5**5

3125

Pero si el mismo valor es asignado a una variable, entonces el resultado no se imprime:

In [4]:
x = 5**5

En ese sentido, en el siguiente código podría ser difícil ver si acaso la función `math.sin` devuelve o imprime el valor:

In [5]:
import math
math.sin(2)

0.9092974268256817

Lo que realmente ocurre es: Importamos el módulo `math`, y llamamos a la función `math.sin` con el argumento `2`. La función `math.sin(2)` *devuelve* el valor `0.909...`, no lo imprmime. Sin embargo, como no hemos asignado el valor devuelto a ninguna variable, la consola Python imprime automáticamente el objeto devuelto.

Dicho de otro modo, el código anterior es una forma resumida de:

In [6]:
import math
print(math.sin(2))

0.9092974268256817


Si en cambio asignamos el valor devuelto a una variable `x`, no se imprime el valor, sino que queda almacenado en la variable `x`:

In [7]:
import math
x = math.sin(2)

Luego podemos operar y/o imprimir la variable `x`:

In [8]:
print(x + 1)

1.9092974268256817


Por lo general las funciones debiesen ejecutarse de manera "silenciosa" (*i.e.* sin imprimir nada) y dar los resultados de su cálculo devolviendo valores.


## Definiendo nuestras propias funciones

En Python podemos programar nuestras propias funciones. El formato genérico para definir una función es el siguiente: 

```python
def miFuncion(arg1, arg2, ..., argn):
    """Texto opcional para la documentación."""

    # Implementación de la función

    return salida  # opcional

# Esto ya no es parte de la función
algun_comando
```

Por ejemplo, la función `bienvenida` definida a continuación imprime el string "¡Hola Mundo!". Es una función sin argumentos de entrada y que no devuelve ningún valor. 

In [9]:
def bienvenida():
    """No recibe nada, solo imprime Hola. """
    print("¡Hola Mundo!")

Si llamamos esta función

In [10]:
bienvenida()

¡Hola Mundo!


In [11]:
type(bienvenida)

function

In [12]:
help(bienvenida)

Help on function bienvenida in module __main__:

bienvenida()
    No recibe nada, solo imprime Hola.



ésta imprime "¡Hola Mundo!", como podría esperarse. 

Asignemos el valor que devuelve esta función a una variable `x`

In [13]:
x = bienvenida()

¡Hola Mundo!


In [14]:
print(x)

None


In [15]:
type(x)

NoneType

In [16]:
otronombre = bienvenida # Para renombrar una función

In [18]:
otronombre()

¡Hola Mundo!


In [20]:
id(otronombre)

140494560020800

In [21]:
id(bienvenida)

140494560020800

encontramos que  `bienvenida` devolvió el objeto `None`. 

Consideremos ahora el siguiente ejemplo:

In [22]:
def muchasBienvenidas(n): 
    print(n * "¡Hola Mundo! ")

En este caso la función `muchasBienvenidas` tiene un argumento `n`. La función imprime el string `¡Hola Mundo! ` repetido `n` veces.

In [23]:
x = muchasBienvenidas(9)

¡Hola Mundo! ¡Hola Mundo! ¡Hola Mundo! ¡Hola Mundo! ¡Hola Mundo! ¡Hola Mundo! ¡Hola Mundo! ¡Hola Mundo! ¡Hola Mundo! 


In [24]:
print(x)

None


Por lo general, las funciones que devuelven valores son más útiles, ya que se pueden combinar con otros códigos, incluyendo otras funciones. Veamos algunos ejemplos de funciones que devuelven un valor.

Supongamos que necesitamos definir una función que calcule el cuadrado de una variable. El código de la función podría ser:

In [25]:
def cuadrado(x):
    """Esta función recibe como argumento x y devuelve x*x"""
    salida = x * x
    return salida  # Aqui devuelve el valor

La palabra clave `def` le dice a Python que en ese punto estamos *definiendo* una función. Esta función toma un argumento (`x`). La función *devuelve* `x * x`, que por supuesto vale $x^2$. 

Esta función puede ser llamada luego en otras partes del código:

In [26]:
cuadrado(10)

100

In [27]:
x = cuadrado(10)

In [28]:
type(x)

int

In [29]:
print(x)

100


In [33]:
help(cuadrado)

Help on function cuadrado in module __main__:

cuadrado(x)
    Esta función recibe como argumento x y devuelve x*x



La variable `salida` usada en la función `cuadrado` se borra una vez que la función devuelve un valor. 

In [32]:
salida

NameError: name 'salida' is not defined

Otro ejemplo

In [34]:
def primeroEnLista(lista):
    """devuelve el primer elemento de una lista. La entrada tiene que ser una lista."""
    x = lista[0]
    return x

In [35]:
primeroEnLista([2,4,6,8,10])

2

In [36]:
listaejemplo = [2,4,6,8,10]
primeroEnLista(listaejemplo)

2

Observar que la función `primeroEnLista` recibe una lista (o secuencia), por lo que si le entregamos otra cosa (por ejemplo un número), obtenemos un error-
.

In [38]:
primeroEnLista(5)

TypeError: 'int' object is not subscriptable

In [39]:
primeroEnLista("Hola")

'H'

In [40]:
def cuadrado(x):
    salida = x * x
    return salida # Hasta aquí definimos la función

# Esta parte no está en la función (está en el programa "principal")
for i in range(5):
    i2 = cuadrado(i)
    print(i, '*', i, '=', i2)
    
x = int(input("Escribe un número: "))
print(cuadrado(x))

0 * 0 = 0
1 * 1 = 1
2 * 2 = 4
3 * 3 = 9
4 * 4 = 16
Escribe un número: 25
625


Observamos que las líneas 1,2,3 definen la función `cuadrado`, mientras las líneas 5-10 son el **programa principal** ("main program")

Podemos definir funciones de más de un argumento:

In [41]:
import math

def hipotenusa(x, y):
    salida = math.sqrt(x * x + y * y)
    return salida

In [42]:
hipotenusa(3,4) # Si doy 3 y 4, me entrega 5 puesto que el triángulo rectangulo de lados 3 y 4 tiene hipotenusa 5

5.0

In [43]:
hipotenusa(3)

TypeError: hipotenusa() missing 1 required positional argument: 'y'

In [44]:
for x in range(1,3):
    for y in range(4,7):
        print("La hipotenusa del triángulo rectángulo de lados", x, "y", y, "vale", hipotenusa(x,y))

La hipotenusa del triángulo rectángulo de lados 1 y 4 vale 4.123105625617661
La hipotenusa del triángulo rectángulo de lados 1 y 5 vale 5.0990195135927845
La hipotenusa del triángulo rectángulo de lados 1 y 6 vale 6.082762530298219
La hipotenusa del triángulo rectángulo de lados 2 y 4 vale 4.47213595499958
La hipotenusa del triángulo rectángulo de lados 2 y 5 vale 5.385164807134504
La hipotenusa del triángulo rectángulo de lados 2 y 6 vale 6.324555320336759


También es posible devolver más de una variable. La función en el siguiente ejemplo recibe un string y devuelve una tupla dos versiones del string, una con todos sus caracteres escritos en mayúsculas y la otra con todos en minúsculas:

In [45]:
def upperAndLower(string):
    return string.upper(), string.lower()

palabra = 'Hola Mundo'

mayus, minus = upperAndLower(palabra)

print(palabra, 'en minúsculas:', minus)
print(palabra, 'en mayúsculas:', mayus)

Hola Mundo en minúsculas: hola mundo
Hola Mundo en mayúsculas: HOLA MUNDO


Se puede definir más de una función en un mismo archivo Python. Por ejemplo:

In [46]:
def muchosAsteriscos( n ):
    return n * '*'

def imprimirConAsteriscos( string ):
    largoMax = 50
    asteriscos = muchosAsteriscos((largoMax - len(string)) // 2)

    print(asteriscos + string + asteriscos)

imprimirConAsteriscos('¡Hola Mundo!')

*******************¡Hola Mundo!*******************


##### Más información

-   [Python Tutorial: Section 4.6 Defining Functions](http://docs.python.org/tutorial/controlflow.html#defining-functions)




-------
### Ejercicio

1. Escriba una función llamada `justificarDerecha` que tenga como argumentos dos strings `texto1` y `texto2` e imprima en dos líneas los strings justificados a la derecha. Por ejemplo:

```
justificarDerecha("Hola como estas?", "bien!")

Hola como estas?
           bien!
```


In [53]:
# Escriba aquí su solución

def justificarDerecha(texto1,texto2):
    # Calculamos el largo de cada palabra
    largo1 = len(texto1)
    largo2 = len(texto2)
    
    # Si la primera palabra es mas larga que la segunda, agregamos (largo1-largo2) espacios al segundo string.
    if(largo1 >= largo2):
        print(texto1)
        print((largo1-largo2) * " " + texto2)
    else:
        print((largo2-largo1) * " " + texto1)
        print(texto2)
        
# Ejemplo donde la primera palaba es más larga que la segunda
justificarDerecha("Hola como estas?", "bien!")

print("")
# Ejemplo donde la segunda palabra es más larga que la primera
justificarDerecha("bien!","Hola como estas?")

Hola como estas?
           bien!

           bien!
Hola como estas?


2. Escriba una función `dibujarGrilla` que reciba dos enteros entero `n` y `k` y dibuje una grilla de `n` por `n` cuadrados, cada uno de lado `k` "guiones". Por ejemplo, `dibujarGrilla(3,4)` imprime:

```
+--------+--------+--------+  
|        |        |        |  
|        |        |        |  
|        |        |        |  
|        |        |        |  
+--------+--------+--------+  
|        |        |        |  
|        |        |        |  
|        |        |        |  
|        |        |        |  
+--------+--------+--------+  
|        |        |        |  
|        |        |        |  
|        |        |        |  
|        |        |        |  
+--------+--------+--------+
```

(fuente: http://greenteapress.com/thinkpython/html/thinkpython004.html#toc39)

In [66]:
# Escriba aquí su solución

def dibujarGrilla(n,k):
    # Definimos primero los dos tipos de línea
    lineaTipo1 = n * ("+" + 2 * k * "-") + "+"
    lineaTipo2 = n * ("|" + 2 * k * " ") + "|"

    for i in range(n):
        print(lineaTipo1)
        for j in range(k):
            print(lineaTipo2)
            
    print(lineaTipo1)

In [68]:
dibujarGrilla(7,3)

+------+------+------+------+------+------+------+
|      |      |      |      |      |      |      |
|      |      |      |      |      |      |      |
|      |      |      |      |      |      |      |
+------+------+------+------+------+------+------+
|      |      |      |      |      |      |      |
|      |      |      |      |      |      |      |
|      |      |      |      |      |      |      |
+------+------+------+------+------+------+------+
|      |      |      |      |      |      |      |
|      |      |      |      |      |      |      |
|      |      |      |      |      |      |      |
+------+------+------+------+------+------+------+
|      |      |      |      |      |      |      |
|      |      |      |      |      |      |      |
|      |      |      |      |      |      |      |
+------+------+------+------+------+------+------+
|      |      |      |      |      |      |      |
|      |      |      |      |      |      |      |
|      |      |      |      |  

3. Un entero se dice creciente si cada uno de sus dígitos es menor que el siguiente. Por ejemplo `578` es creciente y `65` no es creciente. Por convención diremos que un entero positivo menor que `10` es creciente. 
Escriba una función llamada `es_creciente`, que recibe un entero positivo `numero` y devuelve `True` o `False` dependiendo de si `numero` es creciente. 

In [70]:
# Escriba aquí su solución

def es_creciente(n):
    # Primero transformamos el entero en string
    nString = str(n)
    
    # Recorremos los índices
    for i in range(len(nString)-1):
        # Si detectamos un dígito mayor o igual al siguiente
        if (int(nString[i]) >=  int(nString[i+1])):
            # Devolvemos Falso (y la función se detine aquí)
            return False
    # Si nunca se detuvo el ciclo for, concluimos que el entero es creciente y devolvemos Verdadero. 
    return True

In [73]:
es_creciente(5)

True


### Valores por defecto y argumentos opcionales

Python permite definir valores *por defecto* para los argumentos de una función. 

La función del siguiente ejemplo recibe dos argumentos `n`y `hasta`, e imprime la "tabla de multiplicar" de `n`, multiplicando a `n` por todos los enteros entre `1`y `hasta` (incluyendo `hasta`):


In [75]:
def tablaMultiplicar(n, hasta):
    
    print("Tabla del "+str(n)+":")
    for j in range(1,hasta+1):
        print(n, "por", j, "vale", n*j)

tablaMultiplicar(12,17)

Tabla del 12:
12 por 1 vale 12
12 por 2 vale 24
12 por 3 vale 36
12 por 4 vale 48
12 por 5 vale 60
12 por 6 vale 72
12 por 7 vale 84
12 por 8 vale 96
12 por 9 vale 108
12 por 10 vale 120
12 por 11 vale 132
12 por 12 vale 144
12 por 13 vale 156
12 por 14 vale 168
12 por 15 vale 180
12 por 16 vale 192
12 por 17 vale 204


Como normalmente las tablas de multiplicar se enseñan para valores entre 1 y 10, vamos a redefinir la función `tablaMultiplicar` para que asuma por defecto el valor `hasta = 10`:

In [76]:
def tablaMultiplicar2(n, hasta = 10):  # Valor por defecto al argumento "hasta"
    
    print("Tabla del "+str(n)+":")
    for j in range(0,hasta+1):
        print(n, "por", j, "vale", n*j)

De este modo se puede llamar a la función `tablaMultiplicar2` sin necesidad de dar el segundo argumento `hasta`. En ese caso, se asume que la variable `hasta` vale 10:

In [77]:
tablaMultiplicar2(7)

Tabla del 7:
7 por 0 vale 0
7 por 1 vale 7
7 por 2 vale 14
7 por 3 vale 21
7 por 4 vale 28
7 por 5 vale 35
7 por 6 vale 42
7 por 7 vale 49
7 por 8 vale 56
7 por 9 vale 63
7 por 10 vale 70


...y no se pierde nada, en el sentido que el usuario puede asignar otro valor al argumento `hasta` si así lo desea:

In [78]:
tablaMultiplicar2(7,17)

Tabla del 7:
7 por 0 vale 0
7 por 1 vale 7
7 por 2 vale 14
7 por 3 vale 21
7 por 4 vale 28
7 por 5 vale 35
7 por 6 vale 42
7 por 7 vale 49
7 por 8 vale 56
7 por 9 vale 63
7 por 10 vale 70
7 por 11 vale 77
7 por 12 vale 84
7 por 13 vale 91
7 por 14 vale 98
7 por 15 vale 105
7 por 16 vale 112
7 por 17 vale 119


In [82]:
list(range(2,20,2))  # Este es un ejemplo de una función que tiene valores por defecto. 

[2, 4, 6, 8, 10, 12, 14, 16, 18]

In [83]:
help(math.log)  # La función log tiene base e por defecto (logaritmo natural)

Help on built-in function log in module math:

log(...)
    log(x, [base=math.e])
    Return the logarithm of x to the given base.
    
    If the base not specified, returns the natural logarithm (base e) of x.



In [84]:
math.log(5)  # Se asume base e

1.6094379124341003

In [85]:
math.log(5,5) # Se especifica base 5

1.0

### Ejercicio

1. Escriba una función `char_range` que tiene como argumentos tres enteros `inicio`, `fin` y `paso` y entregue un string análogo a `range(inicio, fin, paso)` salvo que con letras en vez de números. Los valores `inicio` y paso `deben` ser opcionales al igual que con `range`. 

Ejemplos:

```python 

>>>char_range(5)
'abcde'

>>> char_range(8,15)
'ijklmno'

>>>char_range(2,20,2)
'cegikmoqs'

```
**Hint:** Puede utilizar la función `chr(n)` que entrega el n-ésimo caracter, de acuerdo a una lista cuyo valor 97 correspodne a la letra "a" y el valor 122 la "z". 



In [89]:
# Escribe aquí tu solución

def char_range(inicio,fin = None,paso = 1):
    
    if(fin == None):
        fin = inicio
        inicio = 0
    
    salida = ""
    for i in range(inicio,fin,paso):
        salida += chr(97+i)
        
    return salida

char_range(0,26,1)

'abcdefghijklmnopqrstuvwxyz'

In [91]:
char_range(26)

'abcdefghijklmnopqrstuvwxyz'

In [93]:
char_range(5,26)

'fghijklmnopqrstuvwxyz'

In [94]:
char_range(0,26,2)

'acegikmoqsuwy'

2.
Escriba una función `decimal_a_base` que tiene como argumentos dos enteros positivos, que llamaremos `n` y `b`. La función debe asumir que `n` está en base decimal, y lo va a transformar a la base `b`. Por lo tanto, la función retorna `numero_b` que corresponde a `n` en base `b`.  Además, el argumento `b` debe ser opcional, teniendo como valor por defecto `2`. 

```python
>>>decimal_a_base(2,2)
10

>>>decimal_a_base(2)
10

>>>decimal_a_base(25)
11001

>>>decimal_a_base(807,4)
30213


```

In [98]:
# Escribe aquí tu solución

def decimal_a_base(n, b = 2):
    
    salida = ""
    
    while(n > 0):
        resto = n % b
        salida = str(resto) + salida
        n = int(n / b) # n // b es la división entera
    
    salida = int(salida)
    
    return(salida)

# Revisamos los ejemplos

print(decimal_a_base(2,2))

print(decimal_a_base(2))

print(decimal_a_base(25))

print(decimal_a_base(807,4))

print(decimal_a_base(373,2))

10
10
11001
30213
101110101


### Modulos

Los modulos (o librerías) permiten agrupar varias funciones en un solo archivo. Normalmente estas funciones tienen una usabilidad común (funciones matemáticas, operaciones vectoriales, cálculo simbólico, graficar, etc). 

Python viene preinstalado con varios módulos, que se pueden conocer con el comando `help('modules')`. En esta sección aprenderemos a utilizar y construir nuestros propios módulos. 


In [99]:
help('modules')


Please wait a moment while I gather a list of all available modules...



/Users/pedromontealegre/opt/anaconda3/lib/python3.9/site-packages/IPython/kernel/__init__.py:12: ShimWarning: The `IPython.kernel` package has been deprecated since IPython 4.0.You should import from ipykernel or jupyter_client instead.
  warn("The `IPython.kernel` package has been deprecated since IPython 4.0."


KeyboardInterrupt: 

### Importando módulos
Como hemos visto anteriormente, para acceder a las funciones de un módulo se utiliza el comando `import`. Por ejemplo, para acceder al módulo de herramientas matemáticas lo hacemos con el comando:

In [100]:
import math

Este comando va a agregar el nombre `math` al *espacio de nombres* (namespace) donde el comando ha sido llamado. Los nombres definidos en el módulo `math` no se agregan directamente, sino que se debe acceder a ellos a través del nombre `math.`. Por ejemplo `math.sin`.

In [101]:
math.sin(1)

0.8414709848078965

In [102]:
math.cos(math.pi)

-1.0

Se puede importar más de un módulo en la misma línea:

In [ ]:
import math, cmath # cmath es una librería de funciones matemáticas para números complejos

Pero la guía de estilos Pyhton [Python Style Guide](http://www.python.org/dev/peps/pep-0008/) recomienda no hacerlo. En cambio, escribimos:

In [ ]:
import math
import cmath 

Se puede escoger un nombre para un módulo distinto a su nombre "oficial".

In [103]:
import math as mathematics

In [104]:
mathematics.sin(1)

0.8414709848078965

In [105]:
import math as mt

In [106]:
mt.sin(1)

0.8414709848078965

Tipicamente esto se usa para:

-   Evitar coincidencias con otros nombres

-   Para cambiar el nombre a algo más manejable. Por ejemplo `import SimpleHTTPServer as shs` 


También se pueden importar objetos individuales de un módulo. Por ejemplo el siguiente código:

In [107]:
from math import sin

Va a importar la función `sin` del módulo `math`. Este código no va a agregar el nombre `math` al espacio de nombres actual, solo el nombre `sin`. 

In [108]:
sin(1)

0.8414709848078965

Es posible importar más de un nombre a la vez:

In [109]:
from math import sin, cos

In [ ]:
cos(0)

o bien todos los nombres del módulo:

In [110]:
from math import *

In [111]:
pi

3.141592653589793

Una vez más, esto no va a agregar el nombre `math` al espacio de nombres actual. Sin embargo va a agregar todos los nombres *publicos* del módulo math al espacio de nombres. En general, es una mala idea hacer esto:

-   Muchos nombres no utilizados se van a guardar inecesariamente en el espacio de nombres. 

-   Probablemente se van a reescribir nombres ya existentes. 

-   Se hará difícil identificar de qué módulo vienen los nombres.

### Creando módulos

Un módulo es en principio simplemente un archivo Python.

Crearemos un ejemplo de un módulo, que guardaremos como `modulo1.py`:

In [115]:
%%file modulo1.py
def unaFuncionMuyUtil(y):
    x = y+1
    return x

palabraMagica = "abracadabra"

Overwriting modulo1.py


Este módulo contiene dos objetos: la función llamada `unaFuncionMuyUtil` y una variable `palabraMagica`.

Podemos ejecutar este archivo (módulo) al igual que cualquier programa Python (por ejemplo escribiendo `python module1.py` en una consola). En este archivo Jupyter escribimos:

In [113]:
!python3 modulo1.py

También podemos importar el módulo:

In [114]:
import modulo1

y así tener acceso a sus objetos:

Cuando Python se encuentra con el comando `import modulo1`, éste busca el archivo `modulo1.py` en el directorio de trabajo actual, y abre el archivo `modulo1.py`. Cuando revisa el archivo `modulo1.py`, Python va a agregar todas las definiciones de funciones y variables del `modulo1` al espacio de nombres al contexto de llamada. Entonces, las variables y funciones definidas en `modulo1.py` van a ser accesibles:

In [117]:
modulo1.unaFuncionMuyUtil(5)

6

In [118]:
modulo1.palabraMagica

'abracadabra'

### La variable  `__name__`

La "variable mágica" `__name__` es una variable especial, predefinida en Python, y que sirve para identificar en qué nivel (de sangría) se está ejecutando el código. La variable toma como valor el string `__main__` si el código es ejecutado en el nivel "0":

In [119]:
__name__

'__main__'

In [120]:
type(__name__)

str

Esta variable se usa cuando se definen módulos. Veamos un ejemplo:

In [123]:
%%file modulo2.py
def unaFuncionMuyUtil(y):
    x = y+1
    return x

print("Soy modulo2, mi variable `__name__` se llama:", __name__)

Overwriting modulo2.py


Si ejecutamos directamente el módulo, obtendremos que la variable `__name__` toma el valor `__main__`:

In [124]:
!python3 modulo2.py   #escribir solo !python si no funciona así

Soy modulo2, mi variable `__name__` se llama: __main__


Por otro lado, si el código es ejecutado como parte de un módulo importado, la variable toma el nombre del archivo que la contiene:

In [1]:
import modulo2      

Soy modulo2, mi variable `__name__` se llama: modulo2


Si ahora escribimos un programa en Python (`programa.py`) que importa el módulo modulo2:

In [3]:
%%file programa1.py

print("Comenzando programa 1")
print("Importamos el modulo 2")
import modulo2

print("Soy programa1, mi variable __name__ se llama:", __name__)

Overwriting programa1.py


In [1]:
!python3 programa1.py  # Reinicien kernel para olvidar que importaron el modulo2

Comenzando programa 1
Importamos el modulo 2
Soy modulo2, mi variable `__name__` se llama: modulo2
Soy programa1, mi variable __name__ se llama: __main__


Vemos que la variable `__name__` toma el valor `modulo2` al ser llamada por `modulo2`, y toma el valor `__main__` cuando quien la llama por `programa1`:

### Utilidad de \_\_name\_\_

En resumen,

-   `__name__` vale `__main__` si el módulo se corre solo (e.g. `python nombreModulo.py`)

-   `__name__` es igual al nombre del módulo (i.e. el nombre sin el sufijo `.py`) si el módulo es importado.

Podemos entonces usar un condicional para escribir código que **solo se ejecuta** cuando el módulo corre solo. Esto es útil para dejar programas de prueba o demostaciones del módulo en este "programa principal condicional". Es una práctica común dejar un programa principal de demostación en cualquier módulo.

In [3]:
%%file modulo3.py
"""Hola"""

def unaFuncionMuyUtil(y):
    x = y+1
    return x

if __name__ == "__main__":
    print("Este modulo contiene una función muy útil.")
    print("Se llama utilizando el comando 'unaFuncionMuyUtil(n)'")
    print("Por ejemplo, si n=5 el valor es:", unaFuncionMuyUtil(5))
else:
    print("En este caso importamos el modulo")

Writing modulo3.py


In [4]:
!python3 modulo3.py

Este modulo contiene una función muy útil.
Se llama utilizando el comando 'unaFuncionMuyUtil(n)'
Por ejemplo, si n=5 el valor es: 6


Si el archivo es importado (*i.e.* usado como módulo), entonces `__name__ == "__main__"` es falso y el bloque condicional no se ejecuta. 

In [7]:
import modulo3

In [ ]:
# Si queremos importar módulos guardados en una carpeta diferente a la "raíz"
# usamos el módulo sys para agregar la dirección de la carpeta al sistema
import sys
sys.path.append('/.../application/app/folder')
# y luego importamos los módulos contenidos en nuestra carpeta como si estuviesen en la raíz
import file_1

#### Más información

-   [Python Tutorial Section 6](http://docs.python.org/tutorial/modules.html#modules)

### Ejercicio


El objetivo de este ejercicio es crear el módulo `bases.py`. Éste módulo debe contener tres funciones:
1. La función `decimal_a_base` que tiene como argumentos dos enteros positivos, que llamaremos `numero` y `base`; y retorna `numero_b` que corresponde a `numero` en base `base`.  El argumento `base` debe ser opcional, teniendo como valor por defecto `2` (esta función fue creada en un ejercicio anterior). 
2. La función `base_a_decimal`, que tiene como argumentos dos enteros positivos, que llamaremos `numero_b` y `base`; y retorna `numero` que corresponde a `numero_b` en base decimal, asumiendo que `numero_b` está en base `base`.  El argumento `base` debe ser opcional, teniendo como valor por defecto `2`.
3. La función `cambio_de_base`, que recibe como argumentos tres enteros positivos `numero`, `base_1` y `base_2` y retorna `numero` en base `base_2`, asumiendo que de entrada el número estaba escrito en `base_1`. Los argumentos `base_1` y `base_2` deben ser opcionales, teniendo como valores por defecto `10` y `2`, respectivamente.

Además, si el módulo es ejecutado solo (no importado en otro programa) entonces debe correr un programa que prueba las tres funciones para valores escogidos por usted. 

(Su programa solo debe funcionar para valores de `base`, `base_1` o `base_2` entre 2 y 9)

In [11]:
%%file bases.py

# Primera función
def decimal_a_base(numero, base = 2):
    salida = ""    
    while(numero > 0):
        resto = numero % base
        salida = str(resto) + salida
        numero = int(numero / base) # n // b es la división entera
    numero_b = int(salida)
    return(numero_b)

#Segunda función
def base_a_decimal(numero_b, base = 2):
    numero_texto = str(numero_b)
    salida = 0
    for i in range(len(numero_texto)):
        salida += int(numero_texto[-(i+1)]) * base ** i
    return(salida)

# Tercera función
def cambio_de_base(numero, base_1 = 10, base_2 = 2):
    numero_decimal = base_a_decimal(numero,base_1)
    salida = decimal_a_base(numero_decimal, base_2)
    return(salida)

if __name__ == "__main__":
    print("Este es 100 en decimal (base 10) escrito en binario (base 2):", decimal_a_base(100))
    print("Este es 100 en binario (base 2) escrito en decimal (base 10):", base_a_decimal(100))
    print("Este es 100 en base 3 escrito en base 7:", cambio_de_base(100,3,7))

Overwriting bases.py


In [12]:
!python3 bases.py

Este es 100 en decimal (base 10) escrito en binario (base 2): 1100100
Este es 100 en binario (base 2) escrito en decimal (base 10): 4
Este es 100 en base 3 escrito en base 7: 12


In [13]:
import bases

In [15]:
bases.decimal_a_base(100)

1100100